index.js

```diff
@@ -1,35 +1,23 @@
 // Public intefrace
 module.exports = promisedPipe
 
+const ensureFunctionType = (fn, i) => {
+    if (typeof fn !== 'function') {
+        throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fn}"`)
+    }
+}
+
+const chain = (q, fn) => q.then(fn)
+
 function promisedPipe(...fns) {
     if (fns.length < 1) {
         throw Error('pipe requires at least one argument')
     }
 
-    fns.forEach((fn, i) => {
-        if (typeof fn !== 'function') {
-            throw Error(`pipe requires each argument to be a function. Argument #${i+1} is of type "${typeof fn}"`)
-        }
-    })
+    fns.forEach(ensureFunctionType)
 
```

Comment:
There are also `.keys` and `.entities` but it seems `babel` transpiles them into an extremely inefficient code.
Well, all-in-all I consider readability a higher priority here.

I find `while` loops very ugly in higher-level programming when you have a lot of other tools to do the same in a more readable way.

The problem of the manual hoisting of the function for `.forEach` is that even having a small file you have to jump with your eyes all around the code just to read it through.

I'm just afraid that your optimization is too micro to count on performance. Note that this code is only executed when you create a pipe but not when you call it.

Maybe let's think of some benchmark so we can run it on Travis to see whether there's any difference?